go/vt/vtgate/vindexes/cfc.go

```diff
@@ -0,0 +1,326 @@
+// functional unique vindexes.
+//
+// This provides the capability to model hierarchical data models. If we
+// consider the prefix as the 'parent' key and the full key as the 'child' key,
+// all the child data is clustered within the same group of shards identified
+// by the 'parent' key.
+//
+// Due to the prevalance of using `vindexes.SingleColumn` in vindexes, it's way
+// more complex to implement a true multi-column composite index (see github
+// issue) than to implement it using a single column vindex where the
+// components of the composite keys are concatenated together to form a single
+// key. The user can use this single key directly as the keyspace id; one can
+// also define a hash function so that the keyspace id is the concatenation of
+// hash(s1), hash(s2), ... hash(sN). Using the concatenated key directly makes
+// it easier to reason the fanout but the data distribution depends on the key
+// itself; while using the hash on components takes care of the randomness of
+// the data distribution.
+//
+// Since the vindex is on a concatenated key, the offsets into the key are the
+// only way to mark its components. Thus it implicitly requires each component
+// to have a fixed width, except the last one. It's especially true when hash
+// is defined. Because the hash is calculated component by component, only the
+// prefix that aligns with the component boundary can be used to compute the
+// key range. Although the misaligned part doesn't participate the key range
+// calculation, the SQL executed on each shard uses the unchanged prefix; thus
+// the behavior is exactly same as other vindex's but just more efficient in
+// controlling the fanout.
+//
+// The expected format of the vindex definition is
+//
+// "vindexes": {
+//   "cfc_md5": {
+//     "type": "cfc",
+//     "params": {
+//       "hash": "md5",
+//       "offsets": "[2,4]"
+//     }
+//   }
+// }
+//
+// 'offsets' only makes sense when hash is used. Offsets should be a sorted
+// list of positive ints, each of which denotes the byte offset (from the
+// beginning of key) of each component's boundary in the concatenated key.
+// Specifically, offsets[0] is the byte offset of the first component,
+// offsets[1] is the byte offset of the second component, etc.
+type CFC struct {
+	name         string
+	hash         func([]byte) []byte
+	offsets      []int
+	prefixVindex SingleColumn
+}
+
+// NewCFC creates a new CFC vindex
+func NewCFC(name string, params map[string]string) (Vindex, error) {
+	// CFC is used in all compare expressions other than 'LIKE'.
+	ss := &CFC{
+		name: name,
+	}
+	// prefixCFC is only used in 'LIKE' compare expressions.
+	ss.prefixVindex = &prefixCFC{name: name, cfc: ss}
+
+	if params == nil {
+		return ss, nil
+	}
+
+	switch h := params["hash"]; h {
+	case "":
+		return ss, nil
+	case "md5":
+		ss.hash = md5hash
+	case "xxhash64":
+		ss.hash = xxhash64
+	default:
+		return nil, fmt.Errorf("invalid hash %s to CFC vindex %s", h, name)
```

Comment:
nit: prefer using `vterrors.Errorf` over `fmt.Errorf`

Reply:
switched to use `vterrors.Errorf`. It's nice that `vterrors.Errorf` has a stacktrace object embedded but it also means `assert.EqualValues` won't work for such kind of errors since the stacktrace is definitely different.